examples/petsc/bps.c

```diff
@@ -385,26 +385,52 @@ static PetscErrorCode Run(RunParams rp, PetscInt num_resources,
 
   PetscFunctionBeginUser;
   // Setup DM
-  if (rp->read_mesh) {
-    ierr = DMPlexCreateFromFile(PETSC_COMM_WORLD, rp->filename, NULL, PETSC_TRUE,
-                                &dm);
-    CHKERRQ(ierr);
-  } else {
-    if (rp->user_l_nodes) {
-      // Find a nicely composite number of elements no less than global nodes
-      PetscMPIInt size;
-      ierr = MPI_Comm_size(rp->comm, &size); CHKERRQ(ierr);
-      for (PetscInt g_elem =
-             PetscMax(1, size * rp->local_nodes / PetscPowInt(rp->degree, rp->dim));
-           ;
-           g_elem++) {
-        Split3(g_elem, rp->mesh_elem, true);
-        if (Max3(rp->mesh_elem) / Min3(rp->mesh_elem) <= 2) break;
-      }
-    }
-    ierr = DMPlexCreateBoxMesh(PETSC_COMM_WORLD, rp->dim, PETSC_FALSE,
-                               rp->mesh_elem,
+  //if (0) {
+  //  if (rp->read_mesh) {
+  //    ierr = DMPlexCreateFromFile(PETSC_COMM_WORLD, rp->filename, NULL, PETSC_TRUE,
+  //                                &dm);
+  //    CHKERRQ(ierr);
+  //  } else {
+  //    if (rp->user_l_nodes) {
+  //      // Find a nicely composite number of elements no less than global nodes
+  //      PetscMPIInt size;
+  //      ierr = MPI_Comm_size(rp->comm, &size); CHKERRQ(ierr);
+  //      for (PetscInt g_elem =
+  //             PetscMax(1, size * rp->local_nodes / PetscPowInt(rp->degree, rp->dim));
+  //           ;
+  //           g_elem++) {
+  //        Split3(g_elem, rp->mesh_elem, true);
+  //        if (Max3(rp->mesh_elem) / Min3(rp->mesh_elem) <= 2) break;
+  //      }
+  //    }
+  //    ierr = DMPlexCreateBoxMesh(PETSC_COMM_WORLD, rp->dim, PETSC_FALSE,
+  //                               rp->mesh_elem,
+  //                               NULL, NULL, NULL, PETSC_TRUE, &dm); CHKERRQ(ierr);
+  //    ierr = DMViewFromOptions(dm, NULL, "-dm_view"); CHKERRQ(ierr);
+  //  }
+  //}
+  // To check if the DMPlexCreateBoxMesh is the issue
+  // Block 1
+  if (1) {
+    // Create DMPLEX
+    ierr = DMCreate(rp->comm, &dm); CHKERRQ(ierr);
+    ierr = DMSetType(dm, DMPLEX); CHKERRQ(ierr);
+    // Set Tensor elements
+    ierr = PetscOptionsSetValue(NULL, "-dm_plex_simplex", "0"); CHKERRQ(ierr);
+    ierr = PetscOptionsSetValue(NULL, "-dm_plex_dim", "3"); CHKERRQ(ierr);
+    ierr = PetscOptionsSetValue(NULL, "-dm_plex_box_faces", "8,8,8"); CHKERRQ(ierr);
+    // Set CL options
+    ierr = DMSetFromOptions(dm); CHKERRQ(ierr);
+    ierr = DMViewFromOptions(dm, NULL, "-dm_view"); CHKERRQ(ierr);
+  }
+  // Block 2
+  if (0) {
+    // Create DMPLEX
+    PetscInt faces[3] = {8,8,8};
+    ierr = DMPlexCreateBoxMesh(rp->comm, rp->dim, PETSC_FALSE, faces,
                                NULL, NULL, NULL, PETSC_TRUE, &dm); CHKERRQ(ierr);
+    ierr = DMSetFromOptions(dm); CHKERRQ(ierr);
+    ierr = DMViewFromOptions(dm, NULL, "-dm_view"); CHKERRQ(ierr);
   }
 
```

Comment:
Thanks @knepley for your reply. To make it simpler, I have 2 blocks that to my knowledge are identical. But if you run with the second block I get in `PetscFEGetDualSpace` function and if I run with the first block, everything is working fine. I am not sure what I am missing!? @jeremylt @jedbrown .

Reply:
I am not good enough with Github to track back from this snippet to the full source, but this is only the DM. Where is the code that adds an FE and runs `DMCreateDS()`?

Reply:
Sorry about that, this is a code
```
  // Block 1
  if (1) {
    // Create DMPLEX
    ierr = DMCreate(rp->comm, &dm); CHKERRQ(ierr);
    ierr = DMSetType(dm, DMPLEX); CHKERRQ(ierr);
    // Set Tensor elements
    ierr = PetscOptionsSetValue(NULL, "-dm_plex_simplex", "0"); CHKERRQ(ierr);
    ierr = PetscOptionsSetValue(NULL, "-dm_plex_dim", "3"); CHKERRQ(ierr);
    ierr = PetscOptionsSetValue(NULL, "-dm_plex_box_faces", "8,8,8"); CHKERRQ(ierr);
    // Set CL options
    ierr = DMSetFromOptions(dm); CHKERRQ(ierr);
    ierr = DMViewFromOptions(dm, NULL, "-dm_view"); CHKERRQ(ierr);
  }
  // Block 2
  if (0) {
    // Create DMPLEX
    PetscInt faces[3] = {8,8,8};
    ierr = DMPlexCreateBoxMesh(rp->comm, rp->dim, PETSC_FALSE, faces,
                               NULL, NULL, NULL, PETSC_TRUE, &dm); CHKERRQ(ierr);
    ierr = DMSetFromOptions(dm); CHKERRQ(ierr);
    ierr = DMViewFromOptions(dm, NULL, "-dm_view"); CHKERRQ(ierr);
  }
```

Reply:
Yes, but you do not set a discretization, `DMAddField()`, anywhere and do not call `DMCreateDS()`.

Reply:
Sorry, I updated above, first I create `fe` for solution, then create `fe_coord`.

Reply:
Yes, that is right.

Reply:
Yeah, when I use first block to create `dm` these lines returns the order of fe equal to `1` (just printed)
```
    ierr = PetscFEGetDualSpace(fe_coord_current, &fe_coord_dual_space);
    CHKERRQ(ierr);
    ierr = PetscDualSpaceGetOrder(fe_coord_dual_space, &fe_coord_order);
    CHKERRQ(ierr);
    printf("order:%d\n", fe_coord_order);
```
But with second block it seems it is not PetscFE. and I get error exactly here.

Reply:
I had hoped that calling DMPlexCreateBoxMesh followed by DMSetFromOptions would setup the coordinate FE, but that doesn't seem to be the case?

We have a lot of logic sitting in a outer loop in this example to generate a lot of different sized DMs with different polynomial orders and numbers of local DoFs so we can benchmark our performance on these different combinations.

I'd like to retain this capability. Years ago we had a shell script call a simplified version of this example that created one DM per each call to the executable, but moving to this code with one executable using the outer loop to create many DMs in the parameter space reduced noise in our benchmark data.

Reply:
And yes, the program looping over all the cases internally is both significantly faster (less startup and mesh setup) and lower noise.

Reply:
Yes, I could not understand what you wanted last night. Jed is correct. With the API path (DMCreateBoxMesh) the intention is that the caller use the API for everything, including creating the coordinate space.

Reply:
Does `DMPlexCreateCoordinateSpace(dm, 1, NULL) ` works for non-tensor case as well?

Reply:
Yes.

Reply:
If I am going to set `degree=1` can I use this one
```
{
    // create FE field for coordinates
    PetscFE fe_coords;
    PetscInt num_comp_coord;
    ierr = DMGetCoordinateDim(dm, &num_comp_coord); CHKERRQ(ierr);
    ierr = PetscFECreateLagrange(comm, dim, num_comp_coord, is_simplex, 1, degree,
                                 &fe_coords); CHKERRQ(ierr);
    ierr = DMProjectCoordinates(dm, fe_coords); CHKERRQ(ierr);
    ierr = PetscFEDestroy(&fe_coords); CHKERRQ(ierr);
  }
```

Reply:
Because the code above works since I don't call `PetscFEGetDualSpace` and I set order to `1` manually.

Reply:
I don't think you need that. @knepley With your logic here, it looks like the degree won't be as specified unless no FE space exists. https://gitlab.com/petsc/petsc/-/blob/main/src/dm/impls/plex/plexcreate.c#L177-197

Inside `DMProjectCoordinates`, there is some code to make a `PetscFE` that merely matches the vertex coordinates. Is there any reason not to have an API that just does this inner work, creating a `PetscFE` that matches the geometry where it lives, without a `DMCreateInterpolation`?
https://gitlab.com/petsc/petsc/-/blob/main/src/dm/interface/dm.c#L6770-6791

Reply:
The code above was originally in `bps.c` with degree set to `1`.

Reply:
@jedbrown Yes, we should add a degree check in the case that FE exists. I would be alright with a refactor exposing the interface that mimics the existing data. We also have to clean up prisms, because that space now works, but I did not find every place to put it in (as in the above link).

Reply:
Do you want a new interface or just `DMProjectCoordinates(dm, NULL)` to create the in-situ `PetscFE` and be done?

Reply:
I only added `is_simplex` case to work for non-tensor basis. I think that is what we want. Now I set order of `fe_coord` to 1 and everything is ok.